### ISSUE TITLE
Allow workers to cancel running task?

### ISSUE THREAD

**Issue by martindurant** [dask/distributed#4694]
Currently, if a task is already running on a thread of a worker, it will always run to completion/exception even if the work is no longer required (e.g., the future was explicitly cancelled). We usually say "you cannot interrupt a python thread", so Dask has no option here. However, you [can](https://blog.finxter.com/how-to-kill-a-thread-in-python/#Method_4_Using_ctypes_To_Raise_Exceptions_In_A_Thread) set the thread state to exception using `ctypes.pythonapi.PyThreadState_SetAsyncExc`. Yes, this is a C-api call, but it is public and stable. Is this worth pursuing? It would allow not only early cancellation of unneeded work, but "interrupts" and timeouts from the worker itself, perhaps even better debugging if it can break deadlocks with a useful traceback.

(Note that the setting of an exception only has effect at the next python bytecode instruction, so some long-running C-level function or even `wait(big)` will not raise immediately)

**Comment by mrocklin** [comment]
If there is a safe and effective way to cancel a thread then I would be
thrilled to see it added to Dask

**Comment by martindurant** [comment]
I'll make a PR with a version of the code linked above, so we can discuss it more there.

**Comment by jrbourbeau** [comment]
Yeah this would be nice to be able to do using a stable, public API

**Comment by mrocklin** [comment]
cc @pitrou who might be able to tell us why this would break things

**Comment by fjetter** [comment]
I have a few concerns
* If the thread spawns subthreads, what happens? Are they closed automagically as well?
* What if the compute thread is supposed to close external resources, e.g. a connection, does it have time to do so, e.g. will the `__exit__` of a ctx manager be executed?

**Comment by martindurant** [comment]
> If the thread spawns subthreads, what happens? 

Whatever would happen on a normal exception, I think. So the working thread, if it was waiting on some other thread, would raise the exception normally. The main thread would not know that there was another thread around, so work may still happen there, unless the code in the task was well-written to pass an "I have reached an exception, please end" message to its child thread.

> if the compute thread is supposed to close external resources, e.g. a connection, does it have time to do so

Yes, it should be the equivalent of meeting the exception at the current location in the code, so contexts would be respected

**Comment by fjetter** [comment]
If that all works this way, that sounds great

**Comment by martindurant** [comment]
That kind of cleanup doesn't happen within a task execution, though, at least not by Dask - these threads only run the functions supplied by the user.

You are right, in general, that code associated with leaving a task would not run if the "kill" happened after areas where an exception might have been anticipated. There's probably no easy way to tell if this is the case for arbitrary user code.

**Comment by pitrou** [comment]
You could do this the other way: expose an API for user code to know whether the current task should be abandoned, and let the user code call it as desired. That allows for safe cancellation. And/or expose something like a `allow_async_cancellation` parameter or function in a task.

**Comment by pitrou** [comment]
Oh, and is it inside a free-running thread or a `ThreadPoolExecutor` thread?

**Comment by martindurant** [comment]
> You could do this the other way: expose an API for user code to know whether the current task should be abandoned, and let the user code call it as desired.

I think it's pretty unlikely we would get any buy-in on this; in fact, you can already do this by introspecting the worker from `get_worker()`.

> is it inside a free-running thread or a ThreadPoolExecutor thread?

The latter, as subclassed in distributed. The exception would be seceded threads (which are still launched the same way). If I remember, an actor thread runs on a executor with exactly one thread.

**Comment by pitrou** [comment]
If your async exception is raised inside the ThreadPoolExecutor bookkeeping code, you might get a deadlock or other issues.

**Comment by martindurant** [comment]
Does any of `ThreadPoolExecutor`'s own code run within the threads it manages?

**Comment by pitrou** [comment]
Well, yes, so that threads are reused. Spawning a new thread for each task would be wasteful.
https://github.com/python/cpython/blob/master/Lib/concurrent/futures/thread.py#L74-L101

**Comment by martindurant** [comment]
Oh, right, I see it.
We can guard this by using our own exception class, right?

**Comment by pitrou** [comment]
I'm not sure that would change anything. You could still interrupt the worker's task handling loop at any point.

**Comment by martindurant** [comment]
I mean like this
(distributed/_concurrent_futures_thread.py)
```python
def _worker(executor_reference, work_queue):
    try:
        while True:
            try:
                work_item = work_queue.get(block=True)
                if work_item is not None:
                    work_item.run()
                    # Delete references to object. See issue16284
                    del work_item
                    continue
                executor = executor_reference()
                # Exit if:
                #   - The interpreter is shutting down OR
                #   - The executor that owns the worker has been collected OR
                #   - The executor that owns the worker has been shutdown.
                if _shutdown or executor is None or executor._shutdown:
                    # Notice other workers
                    work_queue.put(None)
                    return
                del executor
            except WorkerThreadInterrupt:
                # thread interrupt outside of a task
                pass
    except BaseException:
        _base.LOGGER.critical("Exception in worker", exc_info=True)
```

**Comment by pitrou** [comment]
Ah! Well, I suppose that could work...
